dali/operators/segmentation/biased_crop_center.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <random>
+#include <utility>
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/segmentation/utils/searchable_rle_mask.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/core/boundary.h"
+
+#define MASK_SUPPORTED_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, \
+                              uint64_t, int64_t, float)
+
+namespace dali {
+
+DALI_SCHEMA(segmentation__BiasedCropCenter)
+    .DocStr(R"(Selects a cropping window center which can be selected randomly from either
+any position in the input or any position of a foreground pixel in the input, based on an
+``foreground`` argument, typically the result of a coin flip operation.
+
+The purpose of this operator is to obtain a distribution of cropping window centers that
+has a certain bias towards selecting a foreground pixel as a center.
+)")
+    .AddOptionalArg("foreground",
+      R"code(If different than 0, the crop center is selected to match any of the foreground
+ pixels in the input mask. If 0, the crop center is selected randomly.)code",
+      0, true)
+    .AddOptionalArg<std::vector<int>>("shape",
+      R"code(If specified, it represents the shape of the cropping window to be used,
+introducing restrictions to the range of valid crop centers.
+
+When foreground == 0, a random cropping center is selected so that the cropping window
+is fully contained in the input.
+
+When foreground != 0, the cropping center is first picked to match any of the foreground
+pixels in the input. If the selected crop center results in an out of bounds cropping window,
+the center is shifted as necessary so that the window remains within bounds.
```

Comment:
I wonder if it makes any sense not to shift?

Reply:
I think both things can make sense. If cropping out of bounds is OK, the user can let shape=None and any pixel would be considered a valid center. If the shape is specified, we'll assure that the cropping window is within bounds.

Reply:
We've already had github requests for out-of-bounds cropping. Perhaps we should make this configurable.

Reply:
We already have out-of-bounds cropping, and this is already configured (providing a shape or not).